samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartImpl.java

```diff
@@ -16,94 +16,102 @@
 
 package com.example.shoppingcart.domain;
 
-import com.akkaserverless.javasdk.EntityId;
-import com.akkaserverless.javasdk.eventsourcedentity.CommandContext;
-import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntity;
+import com.akkaserverless.javasdk.Effect;
 import com.example.shoppingcart.ShoppingCartApi;
 import com.google.protobuf.Empty;
 
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
-@EventSourcedEntity(entityType = "eventsourced-shopping-cart")
-public class ShoppingCartImpl extends ShoppingCartInterface {
+public class ShoppingCartImpl extends ShoppingCartInterface2 {
   @SuppressWarnings("unused")
   private final String entityId;
 
-  private final Map<String, ShoppingCartApi.LineItem> cart = new LinkedHashMap<>();
-
-  public ShoppingCartImpl(@EntityId String entityId) {
+  public ShoppingCartImpl(String entityId) {
     this.entityId = entityId;
   }
 
   @Override
-  public ShoppingCartDomain.Cart snapshot() {
-    return ShoppingCartDomain.Cart.newBuilder()
-        .addAllItems(cart.values().stream().map(this::convert).collect(Collectors.toList()))
-        .build();
+  protected ShoppingCartDomain.Cart emptyState() {
+    return ShoppingCartDomain.Cart.getDefaultInstance();
   }
 
   @Override
-  public void handleSnapshot(ShoppingCartDomain.Cart cart) {
```

Comment:
One thing we can consider adding would be conversion methods to/from storage representation, so the business logic in command and event handlers can use a more convenient state representation than the proto objects (or json). That class would be somewhat difficult to define in the proto codegen options though, unless it's defined by it's fully qualified java class name.

Reply:
Yeah, working with protobuf objects directly is more awkward. With going to and from builders, changing anything deeper. Would be interesting to see what could be improved, whether that's conversions, or wrappers that make it easier to mutate, in the way lenses are used for this.

Reply:
created issue https://github.com/lightbend/akkaserverless-java-sdk/issues/102 for this idea